microsim/main.py

```diff
@@ -0,0 +1,268 @@
+Core RAMP-UA model.
+
+Created on Wed Apr 29 19:59:25 2020
+
+@author: nick
+"""
+import sys
+
+sys.path.append("microsim")  # This is only needed when testing. I'm so confused about the imports
+import multiprocessing
+import pandas as pd
+
+pd.set_option('display.expand_frame_repr', False)  # Don't wrap lines when displaying DataFrames
+# pd.set_option('display.width', 0)  # Automatically find the best width
+import os
+import click  # command-line interface
+import pickle  # to save data
+from yaml import load, dump, SafeLoader  # pyyaml library for reading the parameters.yml file
+from shutil import copyfile
+
+from microsim.quant_api import QuantRampAPI
+from microsim.population_initialisation import PopulationInitialisation
+from microsim.microsim_model import Microsim
+from microsim.opencl.ramp.run import run_opencl
+from microsim.opencl.ramp.snapshot_convertor import SnapshotConvertor
+from microsim.opencl.ramp.snapshot import Snapshot
+from microsim.opencl.ramp.params import Params
+from microsim.initialisation_cache import InitialisationCache
+
+
+# ********
+# PROGRAM ENTRY POINT
+# Uses 'click' library so that it can be run from the command line
+# ********
+@click.command()
+@click.option('-p', '--parameters_file', default="./model_parameters/default.yml", type=click.Path(exists=True),
+              help="Parameters file to use to configure the model. Default: ./model_parameters/default.yml")
+@click.option('-npf', '--no-parameters-file', is_flag=True,
+              help="Don't read a parameters file, use command line arguments instead")
+@click.option('-i', '--iterations', default=10, help='Number of model iterations. 0 means just run the initialisation')
+@click.option('-s', '--scenario', default="default", help="Name this scenario; output results will be put into a "
+                                                          "directory with this name.")
+@click.option('--data-dir', default="devon_data", help='Root directory to load data from')
+@click.option('--output/--no-output', default=True,
+              help='Whether to generate output data (default yes).')
+@click.option('--output-every-iteration/--no-output-every-iteration', default=False,
+              help='Whether to generate output data at every iteration rather than just at the end (default no).')
+@click.option('--debug/--no-debug', default=False, help="Whether to run some more expensive checks (default no debug)")
+@click.option('-r', '--repetitions', default=1, help="How many times to run the model (default 1)")
+@click.option('-l', '--lockdown-file', default="google_mobility_lockdown_daily.csv",
+              help="Optionally read lockdown mobility data from a file (default use google mobility). To have no "
+                   "lockdown pass an empty string, i.e. --lockdown-file='' ")
+@click.option('--quant-dir', default=None, help='Directory to QUANT data, set to None to use Devon data')
+@click.option('-s', '--use-cache/--dont-use-cache', default=False,
```

Comment:
@rory-improbable I just noticed that some of the short versions of these parameters all share `-s`. Is this delibreate? I'm not sure what the behaviour will be as e.g. how do you distinguish between `@click.option('-s', '--use-cache/--dont-use-cache',` and `@click.option('-s', '--opencl/--no-opencl',` . If this is deliberate it's not a big deal, I can quickly change the duplictes but didn't want to in csae this is intentional.

Reply:
@nickmalleson ahh this is definitely not deliberate, I didn't notice this shorthand option!! If you could change the duplicates to something sensible that would be great

Reply:
good catch, this could definitely have caused some confusion

Reply:
Sounds good, I'll do that now